File: src/algorithms/tspVrp.ts
import { GraphData, AlgorithmExecutionResult, AlgorithmStep } from '../types/graph';
import { runDijkstraSafePath } from './shortestPath';

// Euclidean distance helper
function getDistance(n1: { lat: number; lng: number }, n2: { lat: number; lng: number }): number {
  const dLat = (n2.lat - n1.lat) * 111.32;
  const dLng = (n2.lng - n1.lng) * 111.32 * Math.cos((n1.lat * Math.PI) / 180);
  return Math.sqrt(dLat * dLat + dLng * dLng);
}

// 2-Opt TSP Tour Optimization for Multi-Stop Rescue Vehicles
export function runTspRescueTour(
  graph: GraphData,
  depotNodeId?: string,
  targetNodeIds?: string[]
): AlgorithmExecutionResult {
  const startTime = performance.now();
  const steps: AlgorithmStep[] = [];

  // Determine stops: if targets provided, use them; otherwise pick active distress nodes or residential zones
  const depot = depotNodeId || Object.values(graph.nodes).find(n => n.type === 'hospital' || n.type === 'depot')?.id || Object.keys(graph.nodes)[0];

  let stops = targetNodeIds && targetNodeIds.length > 0
    ? targetNodeIds.filter(id => id !== depot)
    : Object.values(graph.nodes).filter(n => (n.isDistressActive || n.hazardRisk > 0.4) && n.id !== depot).map(n => n.id);

  if (stops.length === 0) {
    stops = Object.keys(graph.nodes).filter(id => id !== depot).slice(0, 6);
  }

  // Tour starts and ends at depot
  const tourNodes = [depot, ...stops];

  steps.push({
    stepIndex: 0,
    description: `TSP Logistics Optimizer initialized for Rescue Ambulance/Supply Convoy. Depot: "${graph.nodes[depot]?.name}". Checkpoints to visit: ${stops.length}.`,
    highlightedNodeIds: tourNodes,
    highlightedEdgeIds: []
  });

  // 1. Initial Tour via Nearest Neighbor heuristic
  const unvisited = new Set(stops);
  const currentTour: string[] = [depot];
  let curr = depot;

  while (unvisited.size > 0) {
    let nearest: string | null = null;
    let minDist = Infinity;

    for (const cand of unvisited) {
      const dist = getDistance(graph.nodes[curr], graph.nodes[cand]);
      if (dist < minDist) {
        minDist = dist;
        nearest = cand;
      }
    }

    if (nearest) {
      currentTour.push(nearest);
      unvisited.delete(nearest);
      curr = nearest;
    }
  }
  // Return to depot
  currentTour.push(depot);

  function calculateTourDistance(tour: string[]): number {
    let d = 0;
    for (let i = 0; i < tour.length - 1; i++) {
      d += getDistance(graph.nodes[tour[i]], graph.nodes[tour[i + 1]]);
    }
    return d;
  }

  let bestDist = calculateTourDistance(currentTour);

  steps.push({
    stepIndex: steps.length,
    description: `Nearest Neighbor initial tour constructed with total direct length ${bestDist.toFixed(1)} km.`,
    highlightedNodeIds: currentTour,
    highlightedEdgeIds: []
  });

  // 2. 2-Opt Optimization Iterations
  let improved = true;
  let iterations = 0;

  while (improved && iterations < 50) {
    improved = false;
    iterations++;

    for (let i = 1; i < currentTour.length - 2; i++) {
      for (let k = i + 1; k < currentTour.length - 1; k++) {
        // Reverse sub-tour from i to k
        const newTour = [
          ...currentTour.slice(0, i),
          ...currentTour.slice(i, k + 1).reverse(),
          ...currentTour.slice(k + 1)
        ];

        const newDist = calculateTourDistance(newTour);
        if (newDist < bestDist - 0.01) {
          bestDist = newDist;
          currentTour.splice(0, currentTour.length, ...newTour);
          improved = true;

          if (steps.length < 30) {
            steps.push({
              stepIndex: steps.length,
              description: `2-Opt Step #${iterations}: Uncrossed path crossing between "${graph.nodes[currentTour[i]]?.name}" and "${graph.nodes[currentTour[k]]?.name}". Tour reduced to ${bestDist.toFixed(1)} km.`,
              highlightedNodeIds: [currentTour[i], currentTour[k]],
              highlightedEdgeIds: []
            });
          }
          break;
        }
      }
      if (improved) break;
    }
  }

  // Trace road-level path along the TSP tour using Dijkstra safe paths
  const allPathEdgeIds: string[] = [];
  for (let i = 0; i < currentTour.length - 1; i++) {
    const res = runDijkstraSafePath(graph, currentTour[i], currentTour[i + 1], true);
    if (res.route) {
      allPathEdgeIds.push(...res.route.pathEdgeIds);
    }
  }

  const tourNames = currentTour.map(id => graph.nodes[id]?.name || id).join(' ➔ ');

  const elapsed = performance.now() - startTime;
  return {
    algorithmName: 'TSP & 2-Opt Multi-Stop Rescue Vehicle Dispatch',
    executionTimeMs: Math.round(elapsed * 100) / 100,
    summary: `Computed optimal minimal-hazard rescue loop visiting ${stops.length} emergency locations. Total tour distance: ${bestDist.toFixed(1)} km.`,
    steps,
    customMetrics: {
      'Stops Visited': stops.length,
      'Total Tour Distance': `${bestDist.toFixed(1)} km`,
      '2-Opt Improvements': iterations,
      'Depot Base': graph.nodes[depot]?.name || depot,
      'Tour Sequence': tourNames
    }
  };
}

export function runCapacitatedVrpDispatch(
  graph: GraphData,
  numVehicles: number = 3,
  vehicleCapacity: number = 150
): AlgorithmExecutionResult {
  const startTime = performance.now();
  const steps: AlgorithmStep[] = [];

  const depot = Object.values(graph.nodes).find(n => n.type === 'hospital' || n.type === 'depot')?.id || Object.keys(graph.nodes)[0];

  const customers = Object.values(graph.nodes)
    .filter(n => n.id !== depot && (n.type === 'residential' || n.isDistressActive || n.hazardRisk > 0.25))
    .map(n => n.id);

  const targetCustomers = customers.length > 0 ? customers : Object.keys(graph.nodes).filter(id => id !== depot).slice(0, 8);

  steps.push({
    stepIndex: 0,
    description: `Capacitated VRP solver initialized with ${numVehicles} rescue convoys (Capacity: ${vehicleCapacity} evacuees) serving ${targetCustomers.length} distress zones from depot "${graph.nodes[depot]?.name}".`,
    highlightedNodeIds: [depot, ...targetCustomers],
    highlightedEdgeIds: []
  });

  const routes: string[][] = Array.from({ length: numVehicles }, () => []);
  const routeLoads: number[] = Array(numVehicles).fill(0);

  const sortedCustomers = [...targetCustomers].sort((a, b) => {
    return (graph.nodes[b]?.hazardRisk || 0) - (graph.nodes[a]?.hazardRisk || 0);
  });

  for (const c of sortedCustomers) {
    const demand = Math.min(vehicleCapacity, Math.max(30, Math.round((graph.nodes[c]?.population || 200) * 0.2)));
    let assigned = false;
    for (let v = 0; v < numVehicles; v++) {
      if (routeLoads[v] + demand <= vehicleCapacity) {
        routes[v].push(c);
        routeLoads[v] += demand;
        assigned = true;
        break;
      }
    }
    if (!assigned) {
      let minV = 0;
      for (let v = 1; v < numVehicles; v++) {
        if (routeLoads[v] < routeLoads[minV]) minV = v;
      }
      routes[minV].push(c);
      routeLoads[minV] += demand;
    }
  }

  let totalFleetDist = 0;
  for (let v = 0; v < numVehicles; v++) {
    if (routes[v].length === 0) continue;
    const fullRoute = [depot, ...routes[v], depot];
    let routeDist = 0;
    for (let i = 0; i < fullRoute.length - 1; i++) {
      routeDist += getDistance(graph.nodes[fullRoute[i]], graph.nodes[fullRoute[i + 1]]);
    }
    totalFleetDist += routeDist;

    const routeNames = fullRoute.map(id => graph.nodes[id]?.name || id).join(' ➔ ');
    steps.push({
      stepIndex: steps.length,
      description: `Convoy #${v + 1} Route (${routeDist.toFixed(1)}km, Load: ${routeLoads[v]}/${vehicleCapacity} evacuees): ${routeNames}`,
      highlightedNodeIds: fullRoute,
      highlightedEdgeIds: []
    });
  }

  const elapsed = performance.now() - startTime;
  const activeRoutes = routes.filter(r => r.length > 0).length;
  const totalServed = routeLoads.reduce((a, b) => a + b, 0);

  return {
    algorithmName: 'Capacitated Vehicle Routing Problem (CVRP Fleet Logistics)',
    executionTimeMs: Math.round(elapsed * 100) / 100,
    summary: `Optimized fleet routes for ${activeRoutes} emergency rescue convoys accommodating ${totalServed} evacuees across ${totalFleetDist.toFixed(1)} km total distance.`,
    steps,
    customMetrics: {
      'Active Rescue Convoys': activeRoutes,
      'Total Fleet Distance': `${totalFleetDist.toFixed(1)} km`,
      'Total Evacuees Accommodated': totalServed,
      'Avg Distance / Vehicle': `${(totalFleetDist / Math.max(activeRoutes, 1)).toFixed(1)} km`,
      'Execution Speed': `${Math.round(elapsed * 100) / 100} ms`
    }
  };
}

File: backend/algorithms/tsp_vrp.py
import math
import time
from typing import Dict, List, Set, Any, Optional
from ..models.graph import GraphData, GraphNode
from ..models.algorithm_results import AlgorithmExecutionResult, AlgorithmStep

def get_distance(node_a: GraphNode, node_b: GraphNode) -> float:
    d_lat = (node_b.lat - node_a.lat) * 111.32
    d_lng = (node_b.lng - node_a.lng) * 111.32 * math.cos(math.radians(node_a.lat))
    return math.sqrt(d_lat * d_lat + d_lng * d_lng)

def run_tsp_rescue_tour(
    graph: GraphData,
    depot_node_id: Optional[str] = None,
    target_node_ids: Optional[List[str]] = None
) -> AlgorithmExecutionResult:
    start_time = time.perf_counter()
    steps: List[AlgorithmStep] = []

    depot = depot_node_id
    if not depot or depot not in graph.nodes:
        for nid, n in graph.nodes.items():
            if n.type in ('hospital', 'depot', 'shelter'):
                depot = nid
                break
        if not depot:
            depot = list(graph.nodes.keys())[0]

    if target_node_ids:
        stops = [nid for nid in target_node_ids if nid in graph.nodes and nid != depot]
    else:
        stops = [nid for nid, n in graph.nodes.items() if (n.isDistressActive or n.hazardRisk > 0.4) and nid != depot]

    if not stops:
        stops = [nid for nid in graph.nodes if nid != depot][:6]

    tour_nodes = [depot] + stops

    steps.append(AlgorithmStep(
        stepIndex=0,
        description=f"TSP Logistics Optimizer initialized for Rescue Ambulance / Supply Convoy. Base Depot: '{graph.nodes[depot].name}'. Checkpoints: {len(stops)}.",
        highlightedNodeIds=tour_nodes,
        highlightedEdgeIds=[]
    ))

    # 1. Nearest Neighbor heuristic
    unvisited = set(stops)
    current_tour = [depot]
    curr = depot

    while unvisited:
        nearest = None
        min_d = float('inf')
        for cand in unvisited:
            d = get_distance(graph.nodes[curr], graph.nodes[cand])
            if d < min_d:
                min_d = d
                nearest = cand
        if nearest:
            current_tour.append(nearest)
            unvisited.remove(nearest)
            curr = nearest

    # Return to depot
    current_tour.append(depot)

    def calc_tour_dist(tour: List[str]) -> float:
        d = 0.0
        for i in range(len(tour) - 1):
            d += get_distance(graph.nodes[tour[i]], graph.nodes[tour[i + 1]])
        return d

    best_dist = calc_tour_dist(current_tour)

    steps.append(AlgorithmStep(
        stepIndex=len(steps),
        description=f"Nearest Neighbor initial tour constructed with total length {best_dist:.1f} km.",
        highlightedNodeIds=current_tour,
        highlightedEdgeIds=[]
    ))

    # 2. 2-Opt improvement
    improved = True
    iterations = 0
    while improved and iterations < 50:
        improved = False
        iterations += 1
        for i in range(1, len(current_tour) - 2):
            for k in range(i + 1, len(current_tour) - 1):
                new_tour = current_tour[:i] + list(reversed(current_tour[i:k + 1])) + current_tour[k + 1:]
                new_dist = calc_tour_dist(new_tour)
                if new_dist < best_dist - 0.01:
                    best_dist = new_dist
                    current_tour = new_tour
                    improved = True
                    if len(steps) < 30:
                        steps.append(AlgorithmStep(
                            stepIndex=len(steps),
                            description=f"2-Opt Step #{iterations}: Uncrossed crossing between '{graph.nodes[current_tour[i]].name}' and '{graph.nodes[current_tour[k]].name}'. Reduced to {best_dist:.1f} km.",
                            highlightedNodeIds=[current_tour[i], current_tour[k]],
                            highlightedEdgeIds=[]
                        ))
                    break
            if improved:
                break

    tour_sequence = " ➔ ".join(graph.nodes[nid].name for nid in current_tour)
    elapsed = (time.perf_counter() - start_time) * 1000

    return AlgorithmExecutionResult(
        algorithmName='TSP & 2-Opt Multi-Stop Rescue Vehicle Dispatch (Python)',
        executionTimeMs=round(elapsed, 2),
        summary=f"Computed optimal minimal-hazard rescue loop visiting {len(stops)} emergency sites. Total tour distance: {best_dist:.1f} km.",
        steps=steps,
        customMetrics={
            'Checkpoints Visited': len(stops),
            'Total Tour Distance': f"{best_dist:.1f} km",
            '2-Opt Improvements': iterations,
            'Depot Base': graph.nodes[depot].name,
            'Tour Sequence': tour_sequence,
            'Execution Speed': f"{round(elapsed, 2)} ms"
        }
    )

def run_capacitated_vrp_dispatch(
    graph: GraphData,
    num_vehicles: int = 3,
    vehicle_capacity: int = 150
) -> AlgorithmExecutionResult:
    start_time = time.perf_counter()
    steps: List[AlgorithmStep] = []

    # Find depot
    depot = None
    for nid, n in graph.nodes.items():
        if n.type in ('hospital', 'depot'):
            depot = nid
            break
    if not depot:
        depot = list(graph.nodes.keys())[0]

    # Customers with demands (endangered communities)
    customers = [
        nid for nid, n in graph.nodes.items()
        if nid != depot and (n.type == 'residential' or n.isDistressActive or n.hazardRisk > 0.25)
    ]
    if not customers:
        customers = [nid for nid in graph.nodes if nid != depot][:8]

    steps.append(AlgorithmStep(
        stepIndex=0,
        description=f"Capacitated VRP solver initialized with {num_vehicles} rescue convoys (Capacity: {vehicle_capacity} evacuees/vehicle) serving {len(customers)} distress zones from depot '{graph.nodes[depot].name}'.",
        highlightedNodeIds=[depot] + customers,
        highlightedEdgeIds=[]
    ))

    # Cluster / Clarke-Wright Savings / Greedy split into routes
    routes: List[List[str]] = [[] for _ in range(num_vehicles)]
    route_loads: List[int] = [0 for _ in range(num_vehicles)]

    sorted_customers = sorted(
        customers,
        key=lambda c: (graph.nodes[c].hazardRisk, graph.nodes[c].population),
        reverse=True
    )

    for c in sorted_customers:
        demand = min(vehicle_capacity, max(30, int(graph.nodes[c].population * 0.2)))
        # Assign to route with least load that has capacity
        assigned = False
        for v_idx in range(num_vehicles):
            if route_loads[v_idx] + demand <= vehicle_capacity:
                routes[v_idx].append(c)
                route_loads[v_idx] += demand
                assigned = True
                break

        if not assigned:
            # Assign to minimum load vehicle
            min_v = min(range(num_vehicles), key=lambda idx: route_loads[idx])
            routes[min_v].append(c)
            route_loads[min_v] += demand

    # Construct complete routes starting & ending at depot
    palette = ['#38bdf8', '#34d399', '#fbbf24', '#f43f5e', '#a855f7']
    total_fleet_dist = 0.0

    for v_idx, route in enumerate(routes):
        if not route:
            continue
        full_route = [depot] + route + [depot]
        route_dist = 0.0
        for i in range(len(full_route) - 1):
            route_dist += get_distance(graph.nodes[full_route[i]], graph.nodes[full_route[i + 1]])
        total_fleet_dist += route_dist

        route_names = " ➔ ".join(graph.nodes[n].name for n in full_route)
        steps.append(AlgorithmStep(
            stepIndex=len(steps),
            description=f"Convoy #{v_idx + 1} Route ({route_dist:.1f}km, Load: {route_loads[v_idx]}/{vehicle_capacity} evacuees): {route_names}",
            highlightedNodeIds=full_route,
            highlightedEdgeIds=[]
        ))

    elapsed = (time.perf_counter() - start_time) * 1000

    active_routes = sum(1 for r in routes if r)
    total_served = sum(route_loads)

    return AlgorithmExecutionResult(
        algorithmName='Capacitated Vehicle Routing Problem (CVRP Fleet Logistics - Python)',
        executionTimeMs=round(elapsed, 2),
        summary=f"Optimized fleet routes for {active_routes} emergency rescue convoys accommodating {total_served} evacuees across {total_fleet_dist:.1f} km total distance.",
        steps=steps,
        customMetrics={
            'Active Rescue Convoys': active_routes,
            'Total Fleet Distance': f"{total_fleet_dist:.1f} km",
            'Total Evacuees Accommodated': total_served,
            'Avg Distance / Vehicle': f"{total_fleet_dist / max(active_routes, 1):.1f} km",
            'Execution Speed': f"{round(elapsed, 2)} ms"
        }
    )
